src/Manager/PicManager.h

```diff
@@ -20,19 +20,37 @@
 
         virtual void grid2par() = 0;
 
-        void setParticleContainer(std::shared_ptr<pc> pcontainer){
+        virtual void dump() = 0;
```

Comment:
1. Some documentation on the pure virtual methods would be useful.
2. Any specific reason the methods are called `grid2par` instead of `gather`? As far as I'm aware, the names of the PIC steps are well established.
3. Is the `dump` method necessary at IPPL core level? Having this as a pure virtual function forces the user to implement it, which means we are assuming things about the user's desired data output features.

Reply:
1. added fde22e6
2. This was suggested in the early template for pic manager by @matt-frey . I actually think grid2par/par2grid are general and self explanatory compared to gather/scatter which may be only known to the pic community.
3. I wanted to put post_step also in AlpineManager to avoid code repetition for mini apps. There, we dump the data at the end of each step. Although the implementation of dump is application specific, the idea of dumping data at the end of each step (or maybe every n steps) is universal among all applications I think and hence it makes sense. @srikrrish @matt-frey what do you think? should I remove it?

Reply:
I think `grid2par` is ok as Mohsen said it is a bit generic than `gather` and hence we can just keep it. I am fine with either moving the `dump` from `PicManager` to `AlpineManager` or having a pure virtual function with a default implementation in `PicManager` itself. But I see that you have changed the pure virtual function to just a virtual function. Is that what @matt-frey you meant with default implementation? because then the class is no more abstract right?

Reply:
To clarify the terminology. If a class has a *pure virtual function*, it is an *abstract class*, i.e. you cannot create an object of this type. You can only declare pure virtual functions, i.e. no function definition. The function definition is provided in derived classes. Hence, pure virtual functions cannot have a default implementation in the abstract class. If you want to provide a default implementation, the function must be `virtual` only.

Reply:
Ok thanks for the clarification.

Reply:
I am also in favor of moving it to the ALPINE manager. The generic PIC manager should only contain symbols that are relevant at library level. Even if the concept of a regular data dump is universal, the exact logic is application-specific. For this purpose, the user is free to override `post_step` as @mohsensadr wrote. In more concrete terms: if the PIC manager never calls the `dump` function, it shouldn't be part of the class; if the symbol is only accessed from user code, it might as well also be defined in user code (also means the user is free to use any name or signature).

Reply:
I moved `dump` to AlpineManger af38feb
Which elements of PicManager/BaseManager should be pure virtual, and which should be virtual? As far as I understand, both can be overwritten, but the user is forced to implement pure virtual ones. Now, everything is virtual. Is it important for BaseManager or PicManager class to be abstract (all member methods/functions to be pure virtual)?

In PicManager

```
	virtual void par2grid() { /* default does nothing */ };
	virtual void grid2par() { /* default does nothing */ };
```

and in BaseManager

```
        virtual void pre_run() { /* default does nothing */}

	virtual void pre_step() { /* default does nothing */}
        
	virtual void post_step() { /* default does nothing */}

	virtual void advance() { /* default does nothing */}
```

Reply:
I would say that, as a general rule, your understanding of the concept also accurately describes the thought process that determines which should be which.

Methods that "do nothing by default" are optional behavior, i.e. the program is correct with or without any nontrivial definition for them. The user is free to ignore them if they are not relevant.

In contrast, a pure virtual function must be implemented, which means that it represents a concept that is required for the program to work but whose exact nature depends on the program. The implementation is application dependent but the concept is general to all applications.

For PIC, the steps in the PIC loop are not optional (otherwise it wouldn't be PIC). These are therefore pure virtual functions, as the PIC cycle always involves each step but the user must decide what happens in those steps. In particular, since the generic PIC manager does not know the particle types, the user must specify which attributes are relevant to the PIC loop[^1]. In contrast, pre-/post-step setup and cleanup are optional behavior; we don't assume that every user needs to have them. Thus, their default implementation is empty.

[^1]: Strictly speaking, there is a way around this by specifying a vector of attributes and associated fields for which the PIC loop is employed (introduces some complications regarding generic types), but I would say that this is beyond the scope of this PR.

Reply:
In 000198c, I made par2grid, grid2par, and advance pure virtual methods. I guess these are the minimum elements of a PIC code. Although, this is very subjective...